File: app/src/main/java/com/electrosim/simulation/SimConnector.java
/**
 * 
 */
package com.electrosim.simulation;

import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;

import com.electrosim.Connector;
import com.electrosim.LinkPointInterface;

/**
 * @author Abdul Hannan
 * 
 */

public class SimConnector extends Connector {

	OutputPort portA;
	InputPort portB;
	Object data;
	boolean flaghigh;

	/**
	 * 
	 */

	Paint paint = new Paint();

	public SimConnector() {

		paint.setColor(Color.RED);
		paint.setStyle(Paint.Style.STROKE);
		paint.setStrokeWidth(3);
		paint.setShadowLayer(2, -1, -1, Color.BLUE);
        paint.setAntiAlias(true);


	}

	@Override
	public boolean setLinkPointA(LinkPointInterface lp) {
		if (lp instanceof OutputPort) {
			this.portA = (OutputPort) lp;
			return super.setLinkPointA(lp);
		}
		return false;
	}

	@Override
	public boolean setLinkPointB(LinkPointInterface lp) {
		if (lp instanceof InputPort) {
			this.portB = (InputPort) lp;

			return super.setLinkPointB(lp);
		}
		return false;
	}

	public void simulate() {

		if (data == null) {
			data = portA.popData();

		}

		// if(data != null)
		// flag=portB.pushData(data);
		// if(flag)
		// data=null;

		if (data != null) {
			if ((Integer) data == 1)
				flaghigh = true;
			else
				flaghigh = false;

			if (portB.pushData(data))
				data = null;

		}

		else
			flaghigh = false;
	}

	protected void prepareToStart() {

		data = null;

	}

	protected void prepareToStop() {

		data = null;
		flaghigh = false;
	}

	@Override
	public void draw(Canvas canvas) {
		if (flaghigh) {
			draw(canvas, paint);
		} else
			super.draw(canvas);
	}

}

File: app/src/main/java/com/electrosim/simulation/SimConnector.kt
/**
 * 
 */
package com.electrosim.simulation

import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import com.electrosim.Connector
import com.electrosim.LinkPointInterface

/**
 * @author Abdul Hannan
 * 
 */
open class SimConnector : Connector() {
    private lateinit var portA: OutputPort
    private lateinit var portB: InputPort
    private var data: Any? = null
    private var flaghigh: Boolean = false

    private val simPaint = Paint().apply {
        color = Color.RED
        style = Paint.Style.STROKE
        strokeWidth = 3f
        setShadowLayer(2f, -1f, -1f, Color.BLUE)
        isAntiAlias = true
    }

    override fun setLinkPointA(lp: LinkPointInterface): Boolean {
        if (lp is OutputPort) {
            this.portA = lp
            return super.setLinkPointA(lp)
        }
        return false
    }

    override fun setLinkPointB(lp: LinkPointInterface): Boolean {
        if (lp is InputPort) {
            this.portB = lp
            return super.setLinkPointB(lp)
        }
        return false
    }

    fun simulate() {
        if (data == null) {
            data = portA.popData()
        }

        if (data != null) {
            flaghigh = (data as? Int) == 1
            
            if (portB.pushData(data)) {
                data = null
            }
        } else {
            flaghigh = false
        }
    }

    fun prepareToStart() {
        data = null
    }

    fun prepareToStop() {
        data = null
        flaghigh = false
    }

    override fun draw(canvas: Canvas) {
        if (flaghigh) {
            super.draw(canvas, simPaint)
        } else {
            super.draw(canvas)
        }
    }
} 